types/src/diem_id_identifier.rs

```diff
@@ -0,0 +1,274 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use serde::{de, Serialize};
+use std::{
+    fmt::{Display, Formatter},
+    str::FromStr,
+};
+
+/// This file implements length and character set limited string types needed
+/// for DiemID as defined by DIP-10: https://github.com/diem/dip/blob/main/dips/dip-10.md
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct DiemId {
+    // A reusable identifier that represents either the source or destination
+    // end-user of a payment transaction. It is unique to per user at VASP level
+    user_identifier: DiemIdUserIdentifier,
+    // A unique string that is mapped to a VASP
+    vasp_domain_identifier: DiemIdVaspDomainIdentifier,
+}
+
+impl DiemId {
+    pub fn new(
+        user_identifier: DiemIdUserIdentifier,
+        vasp_domain_identifier: DiemIdVaspDomainIdentifier,
+    ) -> Self {
+        Self {
+            user_identifier,
+            vasp_domain_identifier,
+        }
+    }
+
+    pub fn new_from_raw(
+        user_identifier: &str,
+        vasp_domain_identifier: &str,
+    ) -> Result<Self, DiemIdParseError> {
+        Ok(DiemId::new(
+            DiemIdUserIdentifier::new(user_identifier)?,
+            DiemIdVaspDomainIdentifier::new(vasp_domain_identifier)?,
+        ))
+    }
+
+    pub fn user_identifier(&self) -> &DiemIdUserIdentifier {
+        &self.user_identifier
+    }
+
+    pub fn vasp_domain_identifier(&self) -> &DiemIdVaspDomainIdentifier {
+        &self.vasp_domain_identifier
+    }
+}
+
+impl Display for DiemId {
+    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
+        write!(
+            f,
+            "{}@{}",
+            self.user_identifier.as_str(),
+            self.vasp_domain_identifier.as_str()
+        )
+    }
+}
+
+impl FromStr for DiemId {
+    type Err = DiemIdParseError;
+
+    fn from_str(s: &str) -> Result<Self, DiemIdParseError> {
+        if !s.contains('@') {
+            return Err(DiemIdParseError::new("DiemId does not have @".to_string()));
+        };
+        let split = s.split('@');
+        let vec = split.collect::<Vec<&str>>();
+        DiemId::new_from_raw(&vec[0], &vec[1])
+    }
```

Comment:
This is close but I think there's a few corner cases that still need to be covered (and adding more negative tests for this would be nice). For example if you have the string `bmwill@novi@foo` this implementation would accept that string, silently dropping the `@foo` part when in reality we'd want to error out on that. You should be able to fix that by using a method like [split_once](https://doc.rust-lang.org/stable/std/primitive.str.html#method.split_once) but since that's not available in rust 1.51 you should be able to use a combination of [find](https://doc.rust-lang.org/stable/std/primitive.str.html#method.find) and [split_at](https://doc.rust-lang.org/stable/std/primitive.str.html#method.split_at).